server/expression/assignment_cast.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"fmt"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/doltgresql/server/functions/framework"
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// AssignmentCast handles assignment casts.
+type AssignmentCast struct {
+	expr     sql.Expression
+	fromType pgtypes.DoltgresType
+	toType   pgtypes.DoltgresType
+}
+
+var _ sql.Expression = (*AssignmentCast)(nil)
+
+// NewAssignmentCast returns a new *AssignmentCast expression.
+func NewAssignmentCast(expr sql.Expression, fromType pgtypes.DoltgresType, toType pgtypes.DoltgresType) *AssignmentCast {
+	return &AssignmentCast{
+		expr:     expr,
+		fromType: fromType,
+		toType:   toType,
+	}
+}
+
+// Children implements the sql.Expression interface.
+func (ac *AssignmentCast) Children() []sql.Expression {
+	return []sql.Expression{ac.expr}
+}
+
+// Eval implements the sql.Expression interface.
+func (ac *AssignmentCast) Eval(ctx *sql.Context, row sql.Row) (any, error) {
+	val, err := ac.expr.Eval(ctx, row)
+	if err != nil || val == nil {
+		return val, err
+	}
+	castFunc := framework.GetAssignmentCast(ac.fromType.BaseID(), ac.toType.BaseID())
+	if castFunc == nil {
+		return nil, fmt.Errorf("target is of type %s but expression is of type %s", ac.toType.String(), ac.fromType.String())
```

Comment:
Error message should note that this there is no type conversion defined, and also echo the offending expression